### ISSUE TITLE
Mouse events cause VRR redraws when direct scanout is inactive

### ISSUE THREAD

**Issue by Gwenodai** [niri-wm/niri#855]
Niri currently doesn't have a way to prevent cursor updates from incurring a frame draw event seperate from those requested by a fullscreen application. This leads to issues with VRR, particularly in games where the mouse is used as an input. For example; in a first person shooter with a framerate cap of 60fps and VRR enabled on a 144Hz monitor, moving the mouse causes the end result rendered framerate to spike up to the max capable refresh rate of the monitor (144Hz) in spite of what the fullscreen game is actually rendering at (60Hz). Obviously the same goes for games where to cursor is visible as well. This also isn't isolated to mouse movements, but also to mouse button events as well, however that only makes a quick 1 frame spike which might make it hard to spot depending on what info your monitor's OSD allows you to see.

I believe this is a common issue with many Wayland compositors, but I was unsure if this was on Niri's radar so I figured I'd add it to the issues list. I'm not sure what the correct way of fixing this issue should be, but I know Hyprland implemented a band-aid fix for it which mostly works. As for if other WM's or DE's have properly fixed it, I have no idea.

I'm honestly unsure if this should be classified as a bug report or maybe an enhancement, but since it causes VRR to function oddly I've reported it as a bug.

### System Information

* niri version: niri 0.1.10 (0.0.git.1533.9824321f)

* Distro: Fedora 41

* GPU: AMD RX 470 + AMD RX 6700 XT

* CPU: AMD Ryzen 9 3900X

**Comment by YaLTeR** [comment]
This is unexpected to me. Niri will try to submit a frame to the monitor only if something changed, and games usually hide the mouse cursor, so if you move the mouse then nothing changes on screen (since it's invisible) and no frame is submitted, so VRR keeps working. And certainly nothing changes when you just click the mouse.

If I test with `mpv` and https://bxt.rs/files/test-videos/60.mp4 then moving the mouse over mpv does break VRR (since the mouse is visible, mpv doesn't hide it when you move the mouse), but moving the mouse over a second monitor keeps VRR working just fine.

Are you sure this is not caused by something else?

**Comment by Gwenodai** [comment]
I am certain of this. This footage is from a fresh Niri install with a barebones config and only the game running. You can see that even though it's an fps with a hidden cursor, there are still extra frames rendered on mouse movement that aren't rendered by the game itself. Around the end of the clip you can see me just spamming left click without moving the mouse. Those spikes are hard to pick up since they're just one frame so my monitor only shows them every few clicks or so.

https://github.com/user-attachments/assets/52d42e09-92e8-4996-a033-27850d7ddde7

**Comment by Gwenodai** [comment]
> moving the mouse over a second monitor keeps VRR working just fine.

That also influences VRR for the other monitor for me btw.

https://github.com/user-attachments/assets/0cb3d1c5-8742-4f92-b4a5-81b540c30523

**Comment by YaLTeR** [comment]
What if you try with mpv and moving the mouse over the other monitor?

**Comment by Gwenodai** [comment]
Same result. Again the quick spikes near the end are the result of clicking the mouse a lot.

https://github.com/user-attachments/assets/5cda0968-299e-44d5-b1b6-dd3f8fe066cd

**Comment by Gwenodai** [comment]
Huh that's odd. There's certainly something going on here for me. This same exact issue (and I do mean completely identical) pops up on Hyprland without enabling the tacked on fix they've implemented to prevent cursor updates when it's active. On the other hand KDE somehow doesn't suffer from this issue at all.

**Comment by Gwenodai** [comment]
Oh that makes sense! I'm running two GPUs with the weaker one rendering everything while the other one handles games. That might prevent my system from directly scanning out, hence causing me to see an issue that otherwise flew under the radar. However I'm not sure why mpv did it as well considering that was running off the same GPU that niri itself was.

**Comment by YaLTeR** [comment]
I was able to reproduce the hidden cursor issue too with Half-Life 1. When it is directly scanned out it's fine, but if I disable direct scanout then it breaks in the same way.

> I'm running two GPUs with the weaker one rendering everything while the other one handles games. That might prevent my system from directly scanning out, hence causing me to see an issue that otherwise flew under the radar. However I'm not sure why mpv did it as well considering that was running off the same GPU that niri itself was.

Yes, cross-GPU scanout is probably disabled (I don't remember the specifics but there are driver bugs with it atm in some cases); not sure why same-GPU doesn't work for you.

**Comment by Sempyos** [comment]
@Gwenodai is this still an issue for you..?

**Comment by Gwenodai** [comment]
> [@Gwenodai](https://github.com/Gwenodai) is this still an issue for you..?

No, not in any of my current setups. I only ran into it on my old dual GPU setup and I'd actually completely forgotten about it. Though I've got no clue if this specific bug still exists or not.

Single GPU wise, I've not had any vrr issues at all that weren't fixed simply by enabling the debug flag merged in https://github.com/niri-wm/niri/pull/1616.

**Comment by Sempyos** [comment]
I feel I'll probably close this for now, and if you or someone else is affected by it again we can reopen
